Initial introduction of Blink Launch Process data into Chromestatus.

jstenback opened the pull request:
Add a set of fields to the entry creation/edit form so that we collect
all the information that is part of the blink launch process. This
also makes chromestatus generate up to date intent templates with the
future intent of eliminating the need to manually create and send
intent emails. This includes making feature entries aware of the
current intent stage they're in as well as holding on to all the
additional data so that creating intent messages for later stages of
the launch process is a matter of updating the entry and generating a
new intent message.

ebidel commented on admin.py:
```diff
@@ -407,7 +440,12 @@ def post(self, path, feature_id=None):
+          intent_to_implement_url=intent_to_implement_url,
+          origin_trial_feedback_url=origin_trial_feedback_url,
+          motivation=self.request.get('motivation'),
+          explainer_links = explainer_links,
```
```suggestion
          explainer_links=explainer_links,
```

ebidel commented on admin.py:
```diff
@@ -418,22 +456,42 @@ def post(self, path, feature_id=None):
+          experiment_timeline=self.request.get('experiment_timeline'),
+          experiment_risks=self.request.get('experiment_risks'),
+          experiment_extension_reason=self.request.get('experiment_extension_reason'),
+          ongoing_constraints = self.request.get('ongoing_constraints')
```
```suggestion
          ongoing_constraints = self.request.get('ongoing_constraints'),
```

ebidel commented on models.py:
```diff
@@ -65,6 +65,23 @@
+  INTENT_EXTEND_TRIAL: 'Extend Origin Trial',
+  INTENT_IMPLEMENT_SHIP: 'Implement and Ship',
+  INTENT_SHIP: 'Ship',
+  INTENT_REMOVE: 'Remove'
```
```suggestion
  INTENT_REMOVE: 'Remove',
```

ebidel commented on models.py:
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
+  motivation = forms.CharField(label='Motivation', required=True, max_length=500,
+      widget=forms.Textarea(attrs={'cols': 50, 'placeholder': 'Explain why the web needs this change. It may be useful to describe what web developers are forced to do without it. When possible, include links to back up your claims in the explainer.', 'maxlength': 500}))
+
+  explainer_links = forms.CharField(label='Explainer link(s)', required=True,
```
Should this be required if, for some cases, only a blink-dev email will have gone out?

ebidel commented on models.py:
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
+      help_text='If your feature was available as an Origin Trial, link to a summary of usage and developer feedback. If not, leave this empty.')
+
+  doc_links = forms.CharField(label='Doc link(s)', required=False, max_length=500,
+      widget=forms.Textarea(attrs={'rows': 4, 'cols': 50, 'placeholder': 'Links to design doc(s), if and when available. [This is not required to send out an intent to implement. Please update the intent thread with the design doc when ready]. An explainer and/or design doc is sufficient to start this process.\n\n[Note: Please include links and data, where possible, to support any claims.]', 'max ...
```
intent to implement -> Intent to Implement

ebidel commented on models.py:
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
+
+  interop_compat_risks = forms.CharField(label='Interoperability and Compatibility Risks', required=True, max_length=500,
+      widget=forms.Textarea(attrs={'cols': 50, 'maxlength': 500}),
+      help_text='Describe the degree of <a target="_blank" href="https://sites.google.com/a/chromium.org/dev/blink?pli=1#TOC-Policy-for-shipping-and-removing-web-platform-API-features">interoperability risk</a>. For a new feature, the main risk is that it fails to become an interoperable part of the web platform if other browsers do not implement it. For a removal, please review our <a target="_b ...
```
Not sure this will work. Do the links render properly in the textarea?

jstenback replied:
Yep, the help text ends up below the actual textarea (as opposed to the "placeholder" text which can not contain markup AFAIK).

ebidel commented on static/js/admin/feature_form.js:
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
+
+let INTENT_IDENTIFIER_NAMES = {};
+
+INTENT_IDENTIFIER_NAMES[INTENT_IMPLEMENT] =      "INTENT_IMPLEMENT";
```
remove the extra alignment spaces throughout.

ebidel commented on static/js/admin/feature_form.js:
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
+INTENT_IDENTIFIER_NAMES[INTENT_IMPLEMENT] =      "INTENT_IMPLEMENT";
+INTENT_IDENTIFIER_NAMES[INTENT_EXPERIMENT] =     "INTENT_EXPERIMENT";
+INTENT_IDENTIFIER_NAMES[INTENT_EXTEND_TRIAL] =   "INTENT_EXTEND_TRIAL";
+INTENT_IDENTIFIER_NAMES[INTENT_IMPLEMENT_SHIP] = "INTENT_IMPLEMENT_SHIP";
```
single quotes

ebidel commented on static/js/admin/feature_form.js:
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
+INTENT_IDENTIFIER_NAMES[INTENT_SHIP] =           "INTENT_SHIP";
+INTENT_IDENTIFIER_NAMES[INTENT_REMOVE] =         "INTENT_REMOVE";
+
+// A object graph mapping form fields to implementation status and whehter or
```
```suggestion
// An object graph mapping form fields to implementation status and whether or
```

ebidel commented on static/js/admin/feature_form.js:
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
+    INTENT_SHIP:           VISIBLE_OPTIONAL,
+    INTENT_REMOVE:         VISIBLE_OPTIONAL
+  },
+}
```
```suggestion
};
```

ebidel commented on static/js/admin/feature_form.js:
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
+}
+
+function intentStageChanged(stage) {
+  function set_required(id, required) {
```
```suggestion
  function setRequired(id, required) {
```

ebidel commented on static/js/admin/feature_form.js:
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
+
+  function hide(id) {
+    // Set the control's table row style to display: none.
+    document.querySelector('#id_' + id).parentNode.parentNode
```
```suggestion
    document.querySelector('#id_' + id).parentNode.parentNode.style.display = 'none';
```

ebidel commented on static/js/admin/feature_form.js:
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
+	.style = "display: none;";
+  }
+
+  let stage_index = INTENT_IDENTIFIER_NAMES[Number(stage.value)];
```
vars should be camelcase

```suggestion
  let stageIndex = INTENT_IDENTIFIER_NAMES[Number(stage.value)];
```

ebidel commented on static/js/admin/feature_form.js:
```diff
@@ -86,6 +75,384 @@ function toggleMilestones(status) {
+    }
+  }
+
+  let intent_stage_name = document.querySelector('#id_intent_stage_name')
```
```suggestion
  let intentStageName = document.querySelector('#id_intent_stage_name');
```

jstenback commented:
I believe the update to this pull request addresses all the feedback thus far, and also takes a stab at fixing the issues the linter pointed out as well. Additional feedback always welcome!

jstenback commented on templates/feature.html:
```diff
@@ -231,6 +231,14 @@ <h3>{% if feature.owner|length == 1 %}Owner{% else %}Owners{% endif %}</h3>
   {% endif %}
 
+  {% if feature.intent_to_implement_url %}
+  <section id="intent_to_implement_url">
```
It is the presence of a long help_text content that makes the table go so wide. I.e. if there's more help_text than the textbox is wide, then the text box width follows the width of the help_text content. Ironically I just went through and moved all the placeholder content out to the help_text to reduce the redundancy, which I think we should do. The alternative would be to move more things into placeholders, which has the unfortunate effect of making the help text invisible when you go back and want to edit a field that already has some content.

Wide textareas seem like a reasonable trade off between the two IMO, but open to other ideas.

jstenback commented on static/js/admin/feature_form.js:
```diff
@@ -19,8 +19,8 @@ form.addEventListener('change', function(e) {
         toggleMilestones(e.target);
-      } else if (e.target.id === 'id_standardization') {
-        toggleSpecLink(e.target);
+      } else if (e.target.id === 'id_intent_stage') {
```
Yep, we ask people to send Intent to Implement messages before we ask them to send Intent to Experiment, so the Intent to Experiment needs a link to the Intent to Implement.

jstenback commented on models.py:
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
+      widget=forms.Textarea(attrs={'cols': 50, 'placeholder': 'Please describe the desired DevTools debugging support for your feature. Consider emailing https://groups.google.com/forum/?fromgroups#!forum/google-chrome-developer-tools to make sure developers are able to debug and inspect code that uses your feature. For new language features in V8 specifically, refer to the debugger support check ...
+      help_text='Description of the desired DevTools debugging support for your feature. Consider emailing the <a href="https://groups.google.com/forum/?fromgroups#!forum/google-chrome-developer-tools">google-chrome-developer-tools</a> list for additional help.')
+
+  all_platforms = forms.BooleanField(required=False, initial=False, label='Supported on all platforms?', help_text='Will this feature be supported on all six Blink platforms (Windows, Mac, Linux, Chrome OS, Android, and Android WebView)?')
```
Yep, good point. I ended up deleting all the placeholder text and consistently (with 4 small exceptions) using help_text instead. That way the text is always visible, even when a user goes back to edit a feature (placeholder text of course disappears as soon as there's any content in a textarea). This is now consistent with other form field types as well, but has the downside of making the form fields wider.

ebidel replied:
The table is set to `width: 100%` right now. You could try this in forms.scss:

```
input[type="text"], input[type="url"], input[type="email"], textarea {
    width: 100%;
    max-width: $max-content-width;
  }
```

and see if that works a little better. Huge inputs look pretty daunting IMO.

jstenback replied:
Tried a few things before I got that to do what I wanted, pushed an update to reflect something that seems reasonable here.

jstenback commented on models.py:
```diff
@@ -892,19 +954,145 @@ class FeatureForm(forms.Form):
+
+  all_platforms = forms.BooleanField(required=False, initial=False, label='Supported on all platforms?', help_text='Will this feature be supported on all six Blink platforms (Windows, Mac, Linux, Chrome OS, Android, and Android WebView)?')
+
+  all_platforms_descr = forms.CharField(label='Platfor Support Explanation', required=False,
```
Fixed.

ebidel commented:
Think this looks good. I'm a little worried about adding more required fields, but I guess the intension of this change is to firm up the process.

ebidel commented:
LMK when you're ready to merge.

jstenback commented:
Merged per in-person discussion with ebidel.
